src/DependencyInjection/DI/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -208,14 +205,26 @@ private ServiceCallSite TryCreateExact(ServiceDescriptor descriptor, Type servic
             return null;
         }
 
-        private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, int slot)
+        private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, int slot, bool throwOnConstraintViolation)
         {
-            if (serviceType.IsConstructedGenericType &&
-                serviceType.GetGenericTypeDefinition() == descriptor.ServiceType)
+            if (serviceType.IsConstructedGenericType && serviceType.GetGenericTypeDefinition() == descriptor.ServiceType)
             {
                 Debug.Assert(descriptor.ImplementationType != null, "descriptor.ImplementationType != null");
                 var lifetime = new ResultCache(descriptor.Lifetime, serviceType, slot);
-                var closedType = descriptor.ImplementationType.MakeGenericType(serviceType.GenericTypeArguments);
+                Type closedType;
+                try
+                {
+                    closedType = descriptor.ImplementationType.MakeGenericType(serviceType.GenericTypeArguments);
+                }
+                catch (ArgumentException ex)
+                {
+                    if (throwOnConstraintViolation)
+                    {
+                        throw new InvalidOperationException(Resources.FormatGenericConstraintViolation(serviceType, descriptor.ImplementationType), ex);
+                    }
+
+                    return null;
```

Comment:
I think this could end up swallowing exceptions that aren't constraint violations but instead just bad registrations. Mismatched arity between the service and implementation type would be an example of this.

Maybe we should make this feature opt-in. What do you think @pakrym?

Reply:
The other way I look at this is - if the type can't be closed today, it's automatic exception and there is zero mitigation as a consumer other than switching containers, so my overall goal was just to stop blowing up if a type couldn't close. It results in edge cases where it can't be closed for other reasons, but that's all I have, an exception type that covers a main failure case and a few much less frequent ones.

Reply:
That seems like an argument for making it optional

Reply:
That seems like an argument for making it optional